### PR TITLE
Add Datomic Cloud support

### PR DESCRIPTION
**Pull request by mdhaney**
I ran into too many problems trying to use the same adapter for both as we discussed, but was able to get Cloud working using its own adapter.

### CODE REVIEW

File: src/main/com/fulcrologic/rad/database_adapters/datomic_cloud.clj [hunk 1]
```diff
@@ -331,38 +331,8 @@
 (def suggested-logging-blacklist
   "A vector containing a list of namespace strings that generate a lot of debug noise when using Datomic. Can
   be added to Timbre's ns-blacklist to reduce logging overhead."
-  ["com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool"
```
Comment (awkay) [inline src/main/com/fulcrologic/rad/database_adapters/datomic_cloud.clj]: Please don't delete things like this.

File: src/main/com/fulcrologic/rad/database_adapters/datomic_cloud.clj [hunk 2]
```diff
@@ -0,0 +1,648 @@
...
+    [com.wsscode.pathom.connect :as pc]
+    [com.wsscode.pathom.core :as p]
+    [datomic.client.api :as d]
+    [edn-query-language.core :as eql]
+    [taoensso.encore :as enc]
+    [compute.datomic-client-memdb.core :as memdb]
+    [taoensso.timbre :as log])
+  (:import (java.util UUID)))
+
+(def type-map
```
Comment (awkay) [inline src/main/com/fulcrologic/rad/database_adapters/datomic_cloud.clj:24]: A *lot* of code duplication in this ns...why not move a lot of this stuff to a Datomic common ns.

**Review by awkay — CHANGES_REQUESTED** [review]

File: deps.edn [hunk 3]
```diff
@@ -8,6 +8,7 @@
            com.wsscode/pathom         {:mvn/version "2.2.30"}
            com.rpl/specter            {:mvn/version "1.1.3"}
            vvvvalvalval/datomock      {:mvn/version "0.2.2"}
+           datomic-client-memdb       {:mvn/version "1.0.2"}
```
Comment (awkay) [inline deps.edn:11]: Is this required everywhere?  Why is this a common dependency?

File: src/test/com/fulcrologic/rad/test_schema_cloud/address.clj [hunk 4]
```diff
@@ -0,0 +1,18 @@
+(ns com.fulcrologic.rad.test-schema-cloud.address
```
Comment (awkay) [inline src/test/com/fulcrologic/rad/test_schema_cloud/address.clj:1]: Was there a real need to create new attrs for tests?  aren't there already attrs for tests?

**Comment by awkay** [comment]
Let me know if you still want to make this contribution. It needs a little cleanup still.

**Comment by awkay** [comment]
I'll reopen this if/when the remaining work is done.

**Comment by tylernisonoff** [comment]
I was experimenting with the PR in hopes of being able to help get it over the finish line

I've gotten a bunch working working on the RAD demo, except the automatic-resolvers aren't working
It seems to be the `pull-many` implementation -- I always get [nil nil ...] back for queries that work with the `d/pull`  if I'm just using one id

I'm using the new [dev-local](https://docs.datomic.com/cloud/dev-local.html) library for local development with cloud , i suppose theres a chance the query is getting evaluated incorrectly there, but not sure. 

I'm fairly new to datomic so still learning complex datalog queries, but will keep experimenting to see if i can get `pull-many` working -- hopefully get can get something working this week

**Comment by awkay** [comment]
The client API (which cloud uses) does not have a pull-many. https://docs.datomic.com/client-api/datomic.client.api.html#var-pull

I think you have to write a wrapper to implement it yourself...I thought that was already there in the PR. I thought the main problem in the PR was that it was kinda messy.

**Comment by tylernisonoff** [comment]
> I think you have to write a wrapper to implement it yourself...I thought that was already there in the PR.

Yes, but the implementation isn't working for me, so I'm gonna try to write one myself / figure out whats going on with the implementation in the PR. 

> I thought the main problem in the PR was that it was kinda messy.

I'll also look into cleaning it up. Maintaining two adapters seems less ideal, I'll look into if theres a way to unify the adapters, and if not, see how else I can clean it up.

**Comment by tylernisonoff** [comment]
Update: Turns out the pull-many implementation is correct! The error was related to my translation of the queries in the demo to a client-compatible syntax. For example:

[In cloud](https://docs.datomic.com/cloud/query/query-data-reference.html#find-specs), only find rel is support for the find-spec  

So I had to translate the demo queries from something like:

```
(defn get-all-categories
  [env query-params]
  (if-let [db (some-> (get-in env [::datomic/databases :production]) deref)]
    (let [ids (d/q '[:find [?id ...]
                     :where
                     [?e :category/label]
                     [?e :category/id ?id]] db)]
      (mapv (fn [id] {:category/id id}) ids))
    (log/error "No database atom for production schema!")))
```

to

```
(defn get-all-categories
  [env query-params]
  (if-let [db (some-> (get-in env [::datomic/databases :production]) deref)]
    (let [ids (d/q '[:find ?id
                     :where
                     [?e :category/label]
                     [?e :category/id ?id]] db)]
      (->> ids
        flatten
        (mapv (fn [id] {:category/id id}))))
    (log/error "No database atom for production schema!")))
```

[Here's a similar PR in the pathom-datomic repo](https://github.com/wilkerlucio/pathom-datomic/pull/1/files)


Given that, this PR is working. I'll look this week into cleaning it up, and seeing if the adapter can be unified.

**Comment by tylernisonoff** [comment]
Hmm, yeah its definitely going to be difficult to get these sharing the same adaptor.

I think the first step would be getting the current adapter to use the client API
["On-Prem supports client or peer access, while Cloud supports only client access."](https://docs.datomic.com/on-prem/moving-to-cloud.html)

the client API does not allow transactors, so not sure what that means for `ensure-transactor-functions!`... 

Alternatively, I can try to migrate most of this to a common namespace, but that will likely break some compatibility (for example, references to `database-adapters.datomic/attribute-schema` would probably go under `database-adapters.datomic-shared/attribute-schema`, or we'd move the adapter to `database-adatpers.datomic-on-prem` and leave `datomic/` for shared stuff).

@awkay if you have advice on how to proceed (namely, if you think we can reasonably modify this logic to fit within the client API or should just refactor to a new ns), let me know! I understand if you don't have time as well.

**Comment by awkay** [comment]
So, there is no need to make it the *same* adapter. I was just hoping that the rote code duplication could be removed. E.g. move anything that can be shared to a common ns, and require it in both. Also, use the main ns for the *keyword* option namespace...there is no need to make diff nses for the keywords that have the same meaning. They can just be in the datomic ns.

**Comment by tylernisonoff** [comment]
>  Also, use the main ns for the keyword option namespace...there is no need to make diff nses for the keywords that have the same meaning. They can just be in the datomic ns.

@awkay  just to make sure I understand (I'm somewhat new to this stuff, so apologies if I have a misunderstanding):

 we want to keep options such as `:com.fulcrologic.rad.database-adapters.datomic/attribute-schema` in the datomic namespace, and use this `.../database-adapaters.datomic` keyspace across both adapters. 

Since theres a lot of shared functionality across the two adapters that makes use of these keywords, these functions would be moved to a common namespace, and we'd have to reference the keywords using the full  path `:com.fulcrologic.rad.database-adapters.datomic/..` 

Is that what you'd like?

Another option I see could be using the common namespace to house these options, which would clean up the internal code here that references it. Additionally, end-users of the library could require the common namespace to make working with these keywords easier. 

Happy to implement whichever you prefer.

**Comment by awkay** [comment]
So, if you look at the other RAD libraries: there are -options nses...i.e. form-options. These are nothing more than defs with docstrings whose values are keywords. Ideally we'd adopt that in this library as well, I just haven't had the time.

So, in terms of the keywords, that's what I'd do for ALL versions of Datomic...one options ns that describes what the keys mean. You don't have to do that, but it would be apprecaited. none of the actualy kws should change from their current fully-qualified name, because that would be a breaking change. Again, look at how I did it in the main RAD library.

The main thing from this PR was what I said about in the inline comments: there are a number of functions that just got copied willy-nilly into a new ns...pure duplication. I'd like those to have some common ns. I realize it can't all be one ns, or we have require problems, and there probably has to be some dynamic require stuff to even make the whole thing workable.

So a `datomic-common` ns seems right.

**Comment by tylernisonoff** [comment]
got it, makes sense 👌

**Comment by mdhaney** [comment]
So I had originally tried to separate the common functions and reuse them, but it's tricky.  The problem is that inevitably all those functions need to use the Datomic API, which differs between client/cloud and peer.  In an earlier iteration of this adapter I tried to solve that problem using protocols, so the common functions are always calling the protocol methods.  You then have a namespace for each datomic client type that implements the protocol using the correct library.  

At some point I removed that layer, although I don't recall exactly why.  It was probably because I was trying to track down other issues and wanted to remove as much complexity as possible until I had the basic functionality correct, and then was time crunched and never added back the abstraction over the datomic libraries.

Just wanted to let you know what you're dealing with and how you might approach it.  Unfortunately I'm not in a position to make the changes myself right now, and it will likely be several months before I have the free time to do that.  But I can try to help answer questions or give tips as much as possible to whoever wants to fix this.

**Comment by tylernisonoff** [comment]
Opened a new PR here: https://github.com/fulcrologic/fulcro-rad-datomic/pull/9

**Comment by awkay** [comment]
closed in favor of #9

**Pull request closed without merging**